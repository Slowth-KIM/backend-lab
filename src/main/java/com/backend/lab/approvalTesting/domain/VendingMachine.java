package com.backend.lab.approvalTesting.domain;

import java.util.*;

public class VendingMachine {

    private String display;
    private List<Integer> coins;
    private Integer balance;
    private String selectedProduct;

    private String dispensedProduct;
    private List<Integer> returns;

    private List<Integer> bank;

    private Map<String, Integer> stock;

    private Map<String, Integer> prices;
    private Integer[] acceptedCoins;



    public VendingMachine(String selectedProduct, HashMap<String, Integer> stock) {
        this("", new ArrayList<>(), 0, selectedProduct, null, new ArrayList<>(), new ArrayList<>(), stock, (HashMap<String, Integer>) Map.of("Cola", 100, "Chips", 50, "Candy", 65), new Integer[]{1, 5, 10, 25});
    }

    public VendingMachine(String display, List<Integer> coins, Integer balance,
                          String selectedProduct, String dispensedProduct, List<Integer> returns, List<Integer> bank,
                          HashMap<String, Integer> stock, HashMap<String, Integer> prices, Integer[] acceptedCoins) {
        this.display = display;
        this.coins = coins;
        this.balance = balance;
        this.selectedProduct = selectedProduct;
        this.dispensedProduct = dispensedProduct;
        this.returns = returns;
        this.bank = bank;
        this.stock = stock;
        this.prices = prices;
        this.acceptedCoins = acceptedCoins;

        displayBalance();
    }

    private void displayBalance() {
        if(this.balance != 0){
            this.display = String.valueOf(this.balance);
        }
        else{
            this.display = "INSERT COIN";
        }
    }

    public void insertCoin(Integer coin){
        if(Arrays.asList(acceptedCoins).contains(coin)){
            this.coins.add(coin);
            this.balance += coin;
            this.displayBalance();
        }
        else{
            this.returns.add(coin);
        }
    }

    public Integer balance() {
        return balance;
    }

    public Integer[] coins() {
        return coins.toArray(new Integer[0]);
    }

    public String display() {
        return display;
    }

    public List<Integer> returns() {
        return returns;
    }

    public List<Integer> returnCoins(){
        this.balance = 0;
        this.returns = this.coins;
        this.coins = new ArrayList<>();
        this.displayBalance();

        return this.returns;
    }
}
