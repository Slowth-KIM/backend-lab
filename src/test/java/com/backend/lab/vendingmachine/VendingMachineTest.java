package com.backend.lab.vendingmachine;

import com.backend.lab.approvalTesting.domain.VendingMachine;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
public class VendingMachineTest {

    private VendingMachine machine;

    private VendingMachinePrinter printer;
    private Map<String, Integer> coins;

    @BeforeEach
    void setUp() {
        machine = new VendingMachine();
        printer = new VendingMachinePrinter(machine);
        coins = new HashMap<String, Integer>(){{
            put("penny", 1);
            put("nickel", 5);
            put("dime", 10);
            put("quarter", 25);
        }};
    }

    @Test
    public void test_accept_coins() {
        // TODO: use the printer and Approvals.verify instead of assertions

        assertEquals("INSERT COIN", machine.display());

        System.out.println(machine.display());
        machine.insertCoin(coins.get("nickel"));

        assertEquals(5, machine.balance().intValue());
        assertArrayEquals(new Integer[]{5}, machine.coins());
        assertEquals("5", machine.display());
    }

    @Test
    public void test_accept_coins_using_printer(){
        machine.insertCoin(coins.get("nickel"));
        Approvals.verify(printer.printEverything());
    }


    @Test
    public void test_return_coins() {
        machine.insertCoin(coins.get("nickel"));
        List<Integer> returns =  machine.returnCoins();

        assertEquals(0, machine.balance().intValue());
        assertEquals(5, returns.get(0).intValue());
    }

    @Test
    public void test_return_coins_using_printer(){
        machine.insertCoin(coins.get("nickel"));
        machine.returnCoins();
        Approvals.verify(printer.printEverything());
    }
}
