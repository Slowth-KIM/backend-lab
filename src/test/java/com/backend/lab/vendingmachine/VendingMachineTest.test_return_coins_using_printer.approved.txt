VendingMachine
Display                                          INSERT COIN
Returns                                                  [5]
Coins                                                     []
Balance                                                    0
